CONFIGURATION.md
# Configuration
In many cases, it can be used without changing the settings because it works with automatic settings by DHCP. <br>
However, if you want to use a static IP address or change the port number to use it as a server, you can change the settings. <br>

Settings are written in the `config.h` file.

## DHCP
```c++
     // When setting the address with DHCP: true
     // If setting a static IP address: false
     constexpr bool use_dhcp = true;
```
Specify `true` for `use_dhcp` to enable DHCP. <br>
To disable DHCP and use a static IP address, set `use_dhcp` to `false`.

## static IP address
```c++
     // static IP address
     namespace static_ip {
         const IPAddress ip_addr(192, 168, 1, 2);
         const IPAddress dns_server(192, 168, 1, 1);
         const IPAddress gateway(192, 168, 1, 1);
         const IPAddress subnet_mask(255, 255, 255, 0);
     }
```
When specifying a static IP address, the following four items must be set.
- `ip_addr` : IP address used by the device
- `dns_server` : DNS server IP address
- `gateway` : IP address of the gateway
- `subnet_mask` : Subnet mask (IP address format)

Only IPv4 addresses can be used for the IP address. <br>
When putting them in `config.h`, put them in parentheses and separate them with commas.

## port number to use for the match
```c++
     // Match listening port
     constexpr uint16_t listen_port = 10023;

     // Default port for match connection
     constexpr uint16_t default_port = 10023;
```
`listen_port` is the port number to listen on. When using it as a server, change it as necessary. <br>
`default_port` is the destination port number to be used if the port number is omitted when specifying the destination.

## debug log output
```c++
     // enable logging
     constexpr bool enable_log = false;

     // log output listening port (for debugging)
     constexpr uint16_t log_listen_port = 23;
```
The debug log output setting is disabled by default. <br>
If enabled, UART transmission wait will occur during log output, which may affect the operation of the game.

Set `enable_log` to `true` to enable logging.
Then, the log is output in the following two ways.
- UART (TX = 1 pin, RX = 2 pin, 115200bps, 8bits, no parity, 1 stop bit)
- TCP port 23 (port number specified in `log_listen_port`)

## MAC address
```c++
     // MAC address
     uint8_t mac_addr[6] = {0x02, 0x20, 0x40, 0x00, 0x00, 0x00};

     // Automatically generate the lower 3-octets of the MAC address based on the Board Unique ID
     constexpr bool use_board_unique_id = true;
```
The MAC address uses a local MAC address starting with `02-20-40`. <br>
In addition, the lower 3-octets address is generated based on the Board Unique ID so that MAC addresses do not overlap when using multiple devices.

## Transmission interval during inactivity
```c++
     // Transmission interval when inactive
     constexpr int report_interval_ms = 40;
```
This is the interval for reporting the modem status information when there is no data to be sent via USB transmission. <br>
Normally you do not need to change.

___Note: This setting is different from the latency timer. ___

In the original ME56PS2, when the data in the USB transmission buffer is less than 1 packet (62 bytes; 64 bytes minus 2 bytes for the header), the packet transmission waits until the latency timer times out. was designed to The initial value of this latency timer is 40ms, which causes transmission delays.

The me56ps2-emulator-rp2040 implementation immediately sends the data in the buffer even if the data in the send buffer is less than one packet. This operation reduces the transfer efficiency of the USB bus, but enables transmission with lower latency.

In other words, me56ps2-emulator-rp2040 does not have a latency timer setting. <br>
Unlike the latency timer, `report_interval_ms` does not delay data transmission, but sets the transmission interval when there is no data to be transmitted.


CONFIGURATION_ja.md
# 設定
多くの場合は、DHCPによる自動設定で動作するため、設定を変更しなくとも利用できます。<br>
しかし、サーバとして利用するため静的IPアドレスを使用する場合や、ポート番号の変更を行う場合は、設定を変更することもできます。<br>

設定は `config.h` ファイルに記述します。

## DHCP
```c++
    // DHCPでアドレスを設定する場合: true
    // 静的IPアドレスを設定する場合: false
    constexpr bool use_dhcp = true;
```
DHCPを有効にする場合、 `use_dhcp` に `true` を指定します。<br>
DHCPを無効化し静的IPアドレスを使用する場合には、 `use_dhcp` に `false` を指定します。

## 静的IPアドレス
```c++
    // 静的IPアドレス
    namespace static_ip {
        const IPAddress ip_addr(192, 168, 1, 2);
        const IPAddress dns_server(192, 168, 1, 1);
        const IPAddress gateway(192, 168, 1, 1);
        const IPAddress subnet_mask(255, 255, 255, 0);
    }
```
静的IPアドレスを指定する場合は、以下の4項目を設定する必要があります。
- `ip_addr` : デバイスが使用するIPアドレス
- `dns_server` : DNSサーバのIPアドレス
- `gateway` : ゲートウェイのIPアドレス
- `subnet_mask` : サブネットマスク (IPアドレス形式)

IPアドレスは、IPv4アドレスのみ利用できます。<br>
`config.h` 中に記載する際は、丸括弧の中にカンマ区切りで記載してください。

## 対戦に使用するポート番号
```c++
    // 対戦用待ち受けポート
    constexpr uint16_t listen_port = 10023;

    // 対戦接続時デフォルトポート
    constexpr uint16_t default_port = 10023;
```
`listen_port` は、待ち受けに使用するポート番号です。サーバとして使用する場合には、必要に応じて変更してください。<br>
`default_port` は、接続先指定時にポート番号を省略した場合に使用する接続先ポート番号です。

## デバッグ用ログ出力
```c++
    // ログ出力を有効にする
    constexpr bool enable_log = false;

    // ログ出力待ち受けポート (デバッグ用)
    constexpr uint16_t log_listen_port = 23;
```
デバッグ用ログ出力の設定は、初期状態では無効化されています。<br>
有効化すると、ログ出力時にUARTの送信待ちが生じ、ゲームの動作に影響が出る場合があります。

ログ出力を有効にするには、 `enable_log` を `true` に設定します。
すると、以下の二通りの方法でログが出力されます。
- UART (TX = 1 pin, RX = 2 pin, 115200bps, 8bits, no parity, 1 stop bit)
- TCP 23番ポート (`log_listen_port` で指定したポート番号)

## MACアドレス
```c++
    // MACアドレス
    uint8_t mac_addr[6] = {0x02, 0x20, 0x40, 0x00, 0x00, 0x00};

    // Board Unique ID を元に MAC アドレスの下位 3-octets を自動生成する
    constexpr bool use_board_unique_id = true;
```
MACアドレスは、 `02-20-40` から始まるロカールMACアドレスを使用しています。<br>
また、複数のデバイスを利用した際にMACアドレスが重複しないよう、Board Unique IDを元に下位 3-octets のアドレスを生成しています。

## 非アクティブ時の通信間隔
```c++
    // 非アクティブ時の通信間隔
    constexpr int report_interval_ms = 40;
```
USB通信で送信すべきデータが無い場合に、モデムのステータス情報を報告する間隔の設定です。<br>
通常は変更する必要はありません。

___注意: この設定は、latency timerとは異なります。___

オリジナルのME56PS2では、USBの送信バッファ中のデータが1パケット分 (64 bytesからヘッダ分の2 bytesを除いた62 bytes) に満たない場合に、latency timerがタイムアウトするまでパケットの送信を待機するように設計されていました。このlatency timerの初期値は40msであり、通信が遅延する原因となります。

me56ps2-emulator-rp2040の実装では、送信バッファ中のデータが1パケット分に満たない場合でも、バッファ中のデータを直ちに送信します。この動作は、USBバスの転送効率の低下を引き起こしますが、より低遅延な通信が実現できます。

つまり、me56ps2-emulator-rp2040にはlatency timerの設定は存在しません。<br>
`report_interval_ms` はlatency timerとは異なり、データ送信の遅延を生じさせるものではなく、送信すべきデータが無い場合の通信間隔の設定です。


config.h
#include <IPAddress.h>

namespace config {
    // DHCPでアドレスを設定する場合: true
    // 静的IPアドレスを設定する場合: false
    constexpr bool use_dhcp = true;

    // 静的IPアドレス
    namespace static_ip {
        const IPAddress ip_addr(192, 168, 1, 2);
        const IPAddress dns_server(192, 168, 1, 1);
        const IPAddress gateway(192, 168, 1, 1);
        const IPAddress subnet_mask(255, 255, 255, 0);
    }

    // 対戦用待ち受けポート
    constexpr uint16_t listen_port = 10023;

    // 対戦接続時デフォルトポート
    constexpr uint16_t default_port = 10023;

    // ログ出力を有効にする
    constexpr bool enable_log = false;

    // ログ出力待ち受けポート (デバッグ用)
    constexpr uint16_t log_listen_port = 23;

    // MACアドレス
    uint8_t mac_addr[6] = {0x02, 0x20, 0x40, 0x00, 0x00, 0x00};

    // Board Unique ID を元に MAC アドレスの下位 3-octets を自動生成する
    constexpr bool use_board_unique_id = true;

    // 非アクティブ時の通信間隔
    constexpr int report_interval_ms = 40;

    // 再送間隔 (0.1ms単位)
    constexpr uint16_t retry_time_value = 2000;

    // 再送回数
    constexpr uint8_t retry_count = 7;
}
